### PR TITLE
[Merged by Bors] - feat: add the refine linter, to flag uses of `refine'`

### PR DESCRIPTION
**Pull request by adomani**
This PR introduces a linter to flag uses of `refine'`, suggesting to use `refine` instead.

`refine` and `refine'` are similar tactics, but they handle meta-variable slightly differently.  This means that they are not completely interchangeable, but `refine` is more readable and (heuristically) tends to be more efficient on average.

This linter is an incentive to discourage uses of `refine'`, without being a ban.

[Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Usage.20of.20refine').

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by adomani** [comment]
!bench

File: Mathlib/Tactic/Linter/RefineLinter.lean [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+/-- The main implementation of the refine linter. -/
+def refineLinter : Linter where run := withSetOptionIn fun _stx => do
+  unless getLinterHash (← getOptions) do
+    return
+  if (← MonadState.get).messages.hasErrors then
+    return
+  let trees ← getInfoTrees
+  for t in trees.toArray do
+    for stx in (refine_tree t) do
+      Linter.logLint linter.refine stx "Please, use `refine` instead of `refine'`!"
```
Comment (jcommelin) [inline Mathlib/Tactic/Linter/RefineLinter.lean:48]: Should this mention that you sometimes need `apply` for performance?

Comment (adomani) [reply Mathlib/Tactic/Linter/RefineLinter.lean]: I added mention of `apply`, thanks!

**Comment by adomani** [comment]
I am benching a simpler version of this linter at #13806.  I left in that PR the `refine'` that this one fixed, as a small scale test that the simpler algorithm still catches the same issues!

**Comment by adomani** [comment]
The linter at #13806 is simpler and caught the same usages as this one: I am closing this PR in favour of the simpler one.

File: Mathlib/Tactic/Linter/RefineLinter.lean [hunk 2]
```diff
@@ -0,0 +1,51 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+#  The "refine" linter
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/RefineLinter.lean:10]:
```suggestion
# The "refine" linter

File: Mathlib/Tactic/Linter/RefineLinter.lean [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+The "refine" linter flags usages of `refine'`.
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/RefineLinter.lean:12]:
```suggestion
The "refine" linter flags usages of the `refine'` tactic.

File: Mathlib/Tactic/Linter/RefineLinter.lean [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace Mathlib.Linter.refine
+
+/-- The refine linter emits a warning on usages of `refine'`. -/
+register_option linter.refine : Bool := {
+  defValue := true
+  descr := "enable the refine linter"
+}
+
+/-- `getRefine' t` returns all usages of `refine'` in the input syntax. -/
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/RefineLinter.lean:31]:
```suggestion
/-- `getRefine' t` returns all usages of the `refine'` tactic in the input syntax `t`. -/

File: Mathlib/Tactic/Linter/RefineLinter.lean [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+def getRefine' : Syntax → Array Syntax
+  | stx@(.node _ kind args) =>
+    let rargs := (args.map getRefine').flatten
+    if kind == ``Lean.Parser.Tactic.refine' then rargs.push stx else rargs
+  | _ => default
+
+/-- Gets the value of the `linter.refine` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.refine o
+
+/-- The main implementation of the refine linter. -/
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/RefineLinter.lean:42]: Would you like to use `@inherit_doc` here, perhaps on an expanded doc-string?

File: Mathlib/Tactic/Linter/RefineLinter.lean [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+def refineLinter : Linter where run := withSetOptionIn fun _stx => do
+  unless getLinterHash (← getOptions) do
+    return
+  if (← MonadState.get).messages.hasErrors then
+    return
+  for stx in (getRefine' _stx) do
+    Linter.logLint linter.refine stx "Please, use `refine` or `apply` instead of `refine'`!"
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/RefineLinter.lean:49]: How about something like "The refine' tactic is discouraged: please strongly consider using `refine` or `apply` instead."

File: Mathlib/Tactic/Linter/RefineLinter.lean [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  defValue := true
+  descr := "enable the refine linter"
+}
+
+/-- `getRefine' t` returns all usages of `refine'` in the input syntax. -/
+partial
+def getRefine' : Syntax → Array Syntax
+  | stx@(.node _ kind args) =>
+    let rargs := (args.map getRefine').flatten
+    if kind == ``Lean.Parser.Tactic.refine' then rargs.push stx else rargs
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/RefineLinter.lean:36]: Let me double-check I understood this correctly. (Up to you if you would like to include some of this as comments.)
- If we have a syntax node with arguments, we recurse into the arguments, finding all `refine'`s there.
- If the current node is a `refine'`, we flag that one also.
- The first arm returns `args` (which is empty when no `refine'` is found).
- `t` is something else, it cannot contain any `refine'`

Comment (grunweg) [reply Mathlib/Tactic/Linter/RefineLinter.lean:36]: I wonder about the last bullet; if this overlooks something. Otherwise, makes sense to me.

**Review by grunweg — COMMENTED** [review]
The implementation looks good to me, with some minor tweaks applied.
You have a merge conflict, and perhaps some further occurrences crept in on master... with all these addressed, feel free to maintainer merge on my behalf.

I'm not sure if zulip already had consensus on landing this weak linter. Perhaps the warning will need rewording... yet, I think this deserves a further look.

File: Mathlib/Tactic/Linter/RefineLinter.lean [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+-/
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+#  The "refine" linter
+
+The "refine" linter flags usages of `refine'`.
+
+The tactics `refine` and `refine'` are similar, but they handle meta-variable slightly differently.
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/RefineLinter.lean:14]:
```suggestion
The tactics `refine` and `refine'` are similar, but they handle meta-variables slightly differently.

File: Mathlib/Tactic/Linter/RefineLinter.lean (hunk 5 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/RefineLinter.lean:42]: I recycled part of the module doc and used `inherit_doc`.  Thanks!

File: Mathlib/Tactic/Linter/RefineLinter.lean (hunk 7 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/RefineLinter.lean:36]: The other constructors for `Syntax` are `.missing`, `.atom` and `.ident`: the `refine'` tactic nests inside it some of those, but they are all subnodes/leaves of a `.node` with kind `Lean.Parser.Tactic.refine'`.  Conclusion: I do not think that this leaves out anything!

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Comment by grunweg** [comment]
For the record/anybody coming here later: this was discussed multiple times on zulip, including
[in late 2023](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/refine.20vs.20refine'),
[in spring 2024](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Usage.20of.20refine')
and [in summer 2024](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Status.20of.20.60refine'.60). The last thread focused more on the final removal of `refine'` and having a linter against it

**Comment by jcommelin** [comment]
bors d+

**Pull request closed without merging**